File: packages/watchmen-analysis-client/src/model/ontology.ts
// ============================================================================
// Virtual Ontology Data Model
// Business (Virtual Object) → Mapping (Link/Field) → Physical (Topic/Table)
//
// Pure type definitions + static UI display configs.
// Factories, helpers, seed data and persistence live in services/ontologyService.ts
// ============================================================================

export type OntologySensitivity = "public" | "internal" | "confidential" | "restricted";

/** A physical table / topic mapped into a virtual object. */
export interface JoinCondition {
	sourceField: string;
	targetField: string;
}

export interface PhysicalTableMapping {
	topicId: string;
	topicName: string;
	/** Optional alias used inside this virtual object, e.g. "cust" for dm_party_customer */
	alias?: string;
	/** Business role of this physical table within the virtual object */
	kind: PhysicalTableKind;
	/** JOIN behaviour when joining to primary: undefined = derive from kind */
	joinType?: PhysicalTableJoinType;
	/** Factor names selected from this physical table that the virtual object exposes */
	fields: string[];
	/** Join from primary table to this table; sourceField is primary field, targetField is current table field. */
	joinConditions?: JoinCondition[];
}

export type PhysicalTableKind = "primary" | "detail" | "profile" | "metric" | "tag" | "lookup";
export type PhysicalTableJoinType = "inner" | "left" | "right" | "full";

/** A link between two virtual objects, resolved via physical table JOIN rules. */
export interface VirtualLink {
	id: string;
	name: string;
	/** Source virtual object id */
	sourceObjectId: string;
	/** Target virtual object id */
	targetObjectId: string;
	/** JOIN type used to resolve the link */
	joinType: "inner" | "left" | "right" | "full";
	/** Join conditions: source physical field → target physical field */
	joinConditions: JoinCondition[];
	/** Optional human-readable description of this link */
	description?: string;
}

/** A derived attribute computed by traversing the virtual link graph. */
export interface DerivedAttribute {
	id: string;
	name: string;
	description?: string;
	/** The virtual object this derived attribute belongs to */
	objectId: string;
	/** Aggregation / arithmetic type */
	aggregate: "count" | "sum" | "avg" | "min" | "max" | "none";
	/** Traversal path: [objectId, linkId, objectId, ...] */
	path: string[];
	/** Final field name used in aggregation */
	targetField?: string;
}

/** A virtual object: the semantic projection of multiple physical tables. */
export interface VirtualObject {
	id: string;
	name: string;
	description: string;
	/** Data source (DataProfile) this virtual object is bound to. */
	datasourceId?: string;
	/** Physical tables projected into this virtual object */
	physicalTables: PhysicalTableMapping[];
	/** Business attributes (field name → source physical table.field) */
	attributes: { name: string; sourceTable: string; sourceField: string }[];
	/** Derived attributes attached to this virtual object */
	derivedAttributes: DerivedAttribute[];
	/** Icon / emoji used in UI */
	icon?: string;
	/** Color used in UI graph */
	color?: string;
}

export interface VirtualOntology {
	id: string;
	/** Backend ontology id (same as id, kept for backend round-trip). */
	ontologyId?: string;
	/** Optimistic-lock version, returned by the backend and required for subsequent saves. */
	version?: number;
	name: string;
	description: string;
	owner: string;
	technicalOwner: string;
	tags: string[];
	sensitivity: OntologySensitivity;
	virtualObjects: VirtualObject[];
	virtualLinks: VirtualLink[];
	createdAt: string;
	updatedAt: string;
}

// ============================================================================
// Static UI display configs
// ============================================================================

export const sensitivityConfig: Record<OntologySensitivity, { label: string; className: string; icon: string }> = {
	public: { label: "Public", className: "bg-green-100 text-green-700", icon: "🌍" },
	internal: { label: "Internal", className: "bg-blue-100 text-blue-700", icon: "🏢" },
	confidential: { label: "Confidential", className: "bg-orange-100 text-orange-700", icon: "🔒" },
	restricted: { label: "Restricted", className: "bg-red-100 text-red-700", icon: "🚨" },
};

export const physicalTableKindConfig: Record<
	PhysicalTableKind,
	{ label: string; className: string; icon: string; defaultJoinType: PhysicalTableJoinType }
> = {
	primary: { label: "Primary", className: "bg-indigo-100 text-indigo-700", icon: "⭐", defaultJoinType: "inner" },
	detail: { label: "Detail", className: "bg-slate-100 text-slate-700", icon: "📎", defaultJoinType: "left" },
	profile: { label: "Profile", className: "bg-blue-100 text-blue-700", icon: "◇", defaultJoinType: "left" },
	metric: { label: "Metric", className: "bg-emerald-100 text-emerald-700", icon: "Σ", defaultJoinType: "left" },
	tag: { label: "Tag", className: "bg-fuchsia-100 text-fuchsia-700", icon: "#", defaultJoinType: "left" },
	lookup: { label: "Lookup", className: "bg-amber-100 text-amber-700", icon: "🔍", defaultJoinType: "inner" },
};

export const physicalTableJoinTypeConfig: Record<PhysicalTableJoinType, { label: string; className: string }> = {
	inner: { label: "INNER", className: "bg-blue-100 text-blue-700" },
	left: { label: "LEFT", className: "bg-emerald-100 text-emerald-700" },
	right: { label: "RIGHT", className: "bg-amber-100 text-amber-700" },
	full: { label: "FULL", className: "bg-purple-100 text-purple-700" },
};

export const joinTypeConfig: Record<VirtualLink["joinType"], { label: string; className: string }> = {
	inner: { label: "INNER JOIN", className: "bg-blue-100 text-blue-700" },
	left: { label: "LEFT JOIN", className: "bg-emerald-100 text-emerald-700" },
	right: { label: "RIGHT JOIN", className: "bg-amber-100 text-amber-700" },
	full: { label: "FULL JOIN", className: "bg-purple-100 text-purple-700" },
};

export const aggregateConfig: Record<DerivedAttribute["aggregate"], { label: string; icon: string }> = {
	count: { label: "COUNT", icon: "#" },
	sum: { label: "SUM", icon: "Σ" },
	avg: { label: "AVG", icon: "μ" },
	min: { label: "MIN", icon: "↓" },
	max: { label: "MAX", icon: "↑" },
	none: { label: "VALUE", icon: "•" },
};

export const virtualObjectColors = [
	"bg-indigo-500",
	"bg-emerald-500",
	"bg-amber-500",
	"bg-rose-500",
	"bg-blue-500",
	"bg-purple-500",
	"bg-cyan-500",
	"bg-orange-500",
];

File: packages/watchmen-model/src/watchmen_model/admin/ontology.py
from enum import Enum
from typing import List, Optional, Union, Dict, Any

from watchmen_utilities import ArrayHelper, ExtendedBaseModel
from watchmen_model.common import TenantBasedTuple, OptimisticLock, TopicId, DataSourceId
from ..common.tuple_ids import OntologyId


class OntologySensitivity(str, Enum):
	PUBLIC = 'public'
	INTERNAL = 'internal'
	CONFIDENTIAL = 'confidential'
	RESTRICTED = 'restricted'


class JoinCondition(ExtendedBaseModel):
	sourceField: Optional[str] = None
	targetField: Optional[str] = None


class FilterCondition(ExtendedBaseModel):
	# 物理字段名，例如 policy_status_code
	field: Optional[str] = None
	# 比较操作符：eq / ne / in / not_in / gt / gte / lt / lte / is_null / is_not_null
	operator: Optional[str] = 'eq'
	# 常量值；单值操作符为标量，列表操作符为列表，空操作符(is_null/is_not_null)忽略
	value: Optional[Union[str, int, float, bool, List[Any]]] = None


class PhysicalTableMapping(ExtendedBaseModel):
	topicId: Optional[TopicId] = None
	topicName: Optional[str] = None
	alias: Optional[str] = None
	# 业务职能：primary / detail / profile / metric / tag / lookup
	kind: Optional[str] = 'detail'
	# JOIN 行为：inner / left / right / full；空则按 kind 推导默认值
	joinType: Optional[str] = None
	fields: Optional[List[str]] = []
	joinConditions: Optional[List[JoinCondition]] = []
	# 行级过滤（key-in 常量），例如 policy_status_code eq "issued"
	filters: Optional[List[FilterCondition]] = []

	def __setattr__(self, name, value):
		if name == 'joinConditions':
			super().__setattr__(name, PhysicalTableMapping._construct_join_conditions(value))
		elif name == 'filters':
			super().__setattr__(name, PhysicalTableMapping._construct_filters(value))
		else:
			super().__setattr__(name, value)

	@staticmethod
	def _construct_join_conditions(values):
		if values is None:
			return []
		return ArrayHelper(values).map(
			lambda x: x if isinstance(x, JoinCondition) else JoinCondition(**x)
		).to_list()

	@staticmethod
	def _construct_filters(values):
		if values is None:
			return []
		return ArrayHelper(values).map(
			lambda x: x if isinstance(x, FilterCondition) else FilterCondition(**x)
		).to_list()


class VirtualObjectAttribute(ExtendedBaseModel):
	name: Optional[str] = None
	sourceTable: Optional[str] = None
	sourceField: Optional[str] = None


class DerivedAttribute(ExtendedBaseModel):
	id: Optional[str] = None
	name: Optional[str] = None
	description: Optional[str] = None
	objectId: Optional[str] = None
	aggregate: Optional[str] = 'count'
	path: Optional[List[str]] = []
	targetField: Optional[str] = None


class VirtualObject(ExtendedBaseModel):
	id: Optional[str] = None
	name: Optional[str] = None
	description: Optional[str] = None
	# 指向业务数据源（watchmen_model.system.DataSource）。
	# 为空时查询将返回空数据（仅 compile-preview 可用）。
	datasourceId: Optional[DataSourceId] = None
	physicalTables: Optional[List[PhysicalTableMapping]] = []
	attributes: Optional[List[VirtualObjectAttribute]] = []
	derivedAttributes: Optional[List[DerivedAttribute]] = []
	icon: Optional[str] = None
	color: Optional[str] = None

	def __setattr__(self, name, value):
		if name == 'physicalTables':
			super().__setattr__(name, VirtualObject._construct_physical_tables(value))
		elif name == 'attributes':
			super().__setattr__(name, VirtualObject._construct_attributes(value))
		elif name == 'derivedAttributes':
			super().__setattr__(name, VirtualObject._construct_derived_attributes(value))
		else:
			super().__setattr__(name, value)

	@staticmethod
	def _construct_physical_tables(values):
		if values is None:
			return []
		return ArrayHelper(values).map(
			lambda x: x if isinstance(x, PhysicalTableMapping) else PhysicalTableMapping(**x)
		).to_list()

	@staticmethod
	def _construct_attributes(values):
		if values is None:
			return []
		return ArrayHelper(values).map(
			lambda x: x if isinstance(x, VirtualObjectAttribute) else VirtualObjectAttribute(**x)
		).to_list()

	@staticmethod
	def _construct_derived_attributes(values):
		if values is None:
			return []
		return ArrayHelper(values).map(
			lambda x: x if isinstance(x, DerivedAttribute) else DerivedAttribute(**x)
		).to_list()


class VirtualLink(ExtendedBaseModel):
	id: Optional[str] = None
	name: Optional[str] = None
	sourceObjectId: Optional[str] = None
	targetObjectId: Optional[str] = None
	joinType: Optional[str] = 'inner'
	joinConditions: Optional[List[JoinCondition]] = []
	description: Optional[str] = None

	def __setattr__(self, name, value):
		if name == 'joinConditions':
			super().__setattr__(name, VirtualLink._construct_join_conditions(value))
		else:
			super().__setattr__(name, value)

	@staticmethod
	def _construct_join_conditions(values):
		if values is None:
			return []
		return ArrayHelper(values).map(
			lambda x: x if isinstance(x, JoinCondition) else JoinCondition(**x)
		).to_list()


class VirtualOntology(ExtendedBaseModel, TenantBasedTuple, OptimisticLock):
	ontologyId: Optional[OntologyId] = None
	name: Optional[str] = None
	description: Optional[str] = None
	owner: Optional[str] = None
	technicalOwner: Optional[str] = None
	tags: Optional[List[str]] = []
	sensitivity: Optional[OntologySensitivity] = OntologySensitivity.INTERNAL
	# dataSourceId 已下放到 VirtualObject.datasourceId。
	virtualObjects: Optional[List[VirtualObject]] = []
	virtualLinks: Optional[List[VirtualLink]] = []

	def __setattr__(self, name, value):
		if name == 'sensitivity':
			super().__setattr__(name, VirtualOntology._construct_sensitivity(value))
		elif name == 'virtualObjects':
			super().__setattr__(name, VirtualOntology._construct_virtual_objects(value))
		elif name == 'virtualLinks':
			super().__setattr__(name, VirtualOntology._construct_virtual_links(value))
		else:
			super().__setattr__(name, value)

	@staticmethod
	def _construct_sensitivity(value):
		if value is None:
			return OntologySensitivity.INTERNAL
		elif isinstance(value, OntologySensitivity):
			return value
		else:
			return OntologySensitivity(value)

	@staticmethod
	def _construct_virtual_objects(values):
		if values is None:
			return []
		return ArrayHelper(values).map(
			lambda x: x if isinstance(x, VirtualObject) else VirtualObject(**x)
		).to_list()

	@staticmethod
	def _construct_virtual_links(values):
		if values is None:
			return []
		return ArrayHelper(values).map(
			lambda x: x if isinstance(x, VirtualLink) else VirtualLink(**x)
		).to_list()
